README.md
# bedrock-angular-form

A [bedrock][] [AngularJS][] module that provides services and directives for
creating and handling forms.  A form can also be dynamically created with a
JSON based layout model.

Uses features from [bedrock-angular-alert][] and [bedrock-angular-resource][].

## Quick Examples

```html
<div ng-controller="TestController as model">
  <form role="form">
    <fieldset>
      <br-input br-model="model.name"
        br-options="{icon: 'tag', name: 'name', label: 'Name'}">
        Please enter your name.
      </br-input>
    </fieldset>
  </form>
</div>
```

```js
// ... TODO ...
```

## Setup

```
bower install bedrock-angular-form
```

Installation of the module followed by a restart of your [bedrock][] server
is sufficient to make the module available to your application.

To manually add **bedrock-angular-form** as a dependency:

```js
angular.module('myapp', ['bedrock.form']);
```

## API

### br-datepicker

Show a date picker form component.

```html
<br-datepicker
  br-model="model.date"
  br-options="{
    format: 'yyyy/MM/dd',
    time: (24*60*60*1000 - 1)
  }"></br-datepicker>
```

### br-form

Display a form UI for a data model based on a layout.  The top level layout
is split into groups built with `br-form-group`.  Each group is composed of
many fields built with `br-form-field`.

```html
<br-form
  br-model="model.data"
  br-groups="model.layout"
  br-options="{
    editable: false
  }"></br-form>
```

See `brFormLibraryService` for how to create the `br-groups` data.

### br-form-field

Used by `br-form-group` to display a field.

TODO

### br-form-group

Used by `br-form` to display a group of fields.

TODO

### br-form-static

Show a properly styled static form field.

```html
<br-form-static
  br-options="{
    inline: true,
    name: 'staticInfo',
    label: 'Static Info'
  }">
  <br-form-static-content>Static content.</br-form-static-content>
  <br-form-static-help>Help details about static content.</br-form-static-help>
</br-form-static>
```

### br-input

Show an input field form component.

```html
<br-input
  br-model="model.label"
  br-options="{
    icon: 'tag',
    name: 'label',
    label: 'Name',
    placeholder: 'Name'
  }">
  Please enter a name.
</br-input>
```

### br-radio-group

Show a radio group form component.

```html
<br-radio-group
  br-model="model.favoriteColor"
  br-group="[
    {label: 'Red', value: 'red'},
    {label: 'Blue', value: 'blue'},
    {label: 'Other', value: 'other'}
  ]"
  br-options="{
    inline: true,
    name: 'favoriteColor',
    label: 'Favorite Color'
  }">
  Please select your favorite color.
</br-radio-group>
```

### br-select

Show a select form component.

TODO: Note about br-display-item,

```html
<br-select
  br-model="model.type"
  br-items="model.types"
  br-display-item="item.label"
  br-options="{
    name: 'label',
    label: 'Type',
    key: 'id'
  }"></br-select>

```

### br-textarea

Show a textarea form component.

```html
<br-textarea
  br-model="model.comment"
  br-options="{
    icon: 'comment',
    name: 'comment',
    label: 'Comment',
    placeholder: 'Comment',
    rows: '3'
  }">
  Please enter a comment.
</br-textarea>
```

### brFormLibraryService

`brFormLibraryService` provides helpers to manage vocabs that can be used to
create form layout data.

- **brFormLibraryService.create()**: Create a new
  `brFormLibraryService.Library`
- **Library.vocabs**: Map of all loaded vocabs indexed by id.
- **Library.hasVocabs**: Flag set if vocabs exist.
- **Library.properties**: Map of all loaded properties indexed by id.
- **Library.groups**: Map of all loaded groups indexed by id.
- **Library.hasGroups**: Flag set if groups exist.
- **Library.graph**: Flattened graph of all properties and groups.
- **Library.load(id)**: Load a vocab URI, expand and compact the data with a
  custom context, reframe in `@link` mode, then find all properties and
  property groups.

One way to use this is to have static context and vocab files on your site
and use `Library.load()` to load the vocab.

## Layout Format

The form layout is created using JSON-LD.  There are a number of custom and
common properties used in a vocab document to define properties and how they
are combined into form groups.

Properties have some common properties and the `rdfs:range` as a type.
Forms use the property types to know what type of field to display.

Propety groups have labels and comments used for display as well as a
ordered list of properties to display with options.

An example context `https://example.com/contexts/example-v1.jsonld`:

```js
{
  "@context": {
    "id": "@id",
    "type": "@type",
    "dc": "http://purl.org/dc/terms/",
    "dc11": "http://purl.org/dc/elements/1.1/",
    "dcterms": "http://purl.org/dc/terms/",
    "owl": "http://www.w3.org/2002/07/owl#",
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
    "schema": "http://schema.org/",
    "xsd": "http://www.w3.org/2001/XMLSchema#",
    ...

    "owl:sameAs": {"@type": "@id"},

    "br": "urn:bedrock:",

    "layout": {"@id": "br:layout", "@type": "@id", "@container": "@list"},
    "property": {"@id": "br:property", "@type": "@id"},
    "propertyGroup": {"@id": "br:propertyGroup", "@type": "@id"},
    "optional": {"@id": "br:optional", "@type": "@xsd:boolean"},
    "resource": {"@id": "br:resource", "@type": "@id"},
    "date": {"@id": "br:date", "@type": "xsd:dateTime"},
    "domain": {"@id": "rdfs:domain",  "@type": "@vocab"},
    "range": {"@id": "rdfs:range", "@type": "@vocab"},
    "rangeOption": {
      "@id": "br:rangeOption",
      "@container": "@list"
    },
    "rangeOptionCompareProperty": {
      "@id": "br:rangeOptionCompareProperty",
      "@type": "@vocab"
    },
    "comment": "rdfs:comment",
    "label": "rdfs:label",
    "value": "rdf:value",
    "Property": "rdf:Property",
    "PropertyGroup": "br:PropertyGroup",
    "URL": "rdfs:Resource",
    "String": "rdfs:Literal",
    "Date": "xsd:dateTime",

    "ex": "https://example.com/schema/",
    "testId": "ex:testId",
    "testDate": {"@id" : "ex:testDate", "@type": "xsd:dateTime"},
    ...
  }
}
```

An example vocab `https://example.com/vocabs/example-v1.jsonld`:

```js
{
  "@context": "../contexts/example-v1.jsonld",
  "id": "https://example.com/vocabs/example-v1",
  "label": "Example Vocabulary",
  "@graph": [
    {
      "id": "https://example.com/schema/testId",
      "type": "Property",
      "label": "Test ID",
      "comment": "A test identifier.",
      "domain": "schema:Thing",
      "range": "String"
    },
    {
      "id": "https://example.com/schema/testDate",
      "type": "Property",
      "label": "Test Date",
      "comment": "A test date.",
      "domain": "schema:Thing",
      "range": "Date"
    },
    ...
    {
      "id": "https://example.com/schema/examplePropertyGroup",
      "type": "PropertyGroup",
      "label": "Example Group",
      "comment": "The custom properties for an example layout group.",
      "layout": [
        {
          "property": "https://example.com/schema/testId"
        },
        {
          "property": "https://example.com/schema/testDate"
        }
      ]
    }
    ...
  ]
}
```

## Constants

- **monthNames**: Names of months: `['January', 'February', ..., 'December']`
- **monthNumbers**: Numbers of months: `['01', '02', ..., '12']`
- **monthLabels**: Month labels suitable for a `br-select`: `[{index: 1, label: '01 - January', ...]`
- **years**: Next 10 years suitable for expiration use: `[2015, 2016, ..., 2024]`
- **countries**: Countries in sorted display order: `[..., {code: 'XY', name: 'Xxx Yyy'}, ...]`

[bedrock]: https://github.com/digitalbazaar/bedrock
[bedrock-angular]: https://github.com/digitalbazaar/bedrock-angular
[bedrock-angular-alert]: https://github.com/digitalbazaar/bedrock-angular-alert
[bedrock-angular-resource]: https://github.com/digitalbazaar/bedrock-angular-resource
[AngularJS]: https://github.com/angular/angular.js


CHANGELOG.md
# bedrock-angular-form ChangeLog

## 4.0.0 - 2018-03-19

### Changed
- **BREAKING**: Switch to material design.
- Remove bootstrap classes from main controls.
- Update datepicker to use material.

## 3.1.2 - 2017-10-23

### Fixed
- Add missing jsonld dependency by way of bedrock-angular-jsonld.

## 3.1.1 - 2017-09-08

### Fixed
- Remove legacy code.

## 3.1.0 - 2017-09-06

### Changed
- Remove single internal use of bedrock-angular-alert to
  avoid need for the dependency. Single use was for a
  configuration or network error that may occur when
  preloading vocabularies in the form library service.
  This is not useful information for a user nor is it
  recoverable; it has been moved to use console.error().

## 3.0.8 - 2017-08-18

### Fixed
- Evaluate `brOptions` in the context of the parent scope.

## 3.0.7 - 2017-08-10

### Fixed
- Use `querySelector` for CSS selector instead of jqLite `.find`
  which only supports tag names.

## 3.0.6 - 2017-06-05

### Fixed
- Remove angular `ui-select` dependency.

## 3.0.5 - 2017-06-05

### Changed
- Remove `ui-select` dependency.

## 3.0.4 - 2017-06-02

### Fixed
- Add angular-material.css.

## 3.0.3 - 2017-05-31

### Fixed
- Ensure all input element types are updated in br-input.

## 3.0.2 - 2017-05-30

### Fixed
- Update deps.

## 3.0.1 - 2017-05-30

### Fixed
- Add `bootstrap` dependency.

## 3.0.0 - 2017-05-26

### Changed
- **BREAKING**: Switch package manager from bower to npm.
- **BREAKING**: Replace requirejs/amd with ES6 import.
- Angular 1.6.x is required.

## 2.13.3 - 2017-05-04

## Fixed
- Set error class on field when validation is visible.
- Ensure `contentElement` is found when setting event handler.

## 2.13.2 - 2017-05-01

## Fixed
- Fix bug that briefly flashed validation input.
- Remove erroneous checking for presence of error html.

## 2.13.1 - 2017-03-16

### Fixed
- Fix the default `compare` function in the `brSelect` component.

## 2.13.0 - 2017-03-10

### Added
- Add `afterLabel` help toggle positioning option.

## 2.12.1 - 2017-02-15

### Fixed
- Fix bug where options are not specified as a literal.

## 2.12.0 - 2017-02-06

### Added
- Show validation errors by default in textarea directive.
- Add default validation text in textarea directive.

## 2.11.1 - 2016-12-01

### Fixed
- Remove obsolete `version` tag from bower.json.

## 2.11.0 - 2016-12-01

### Added
- Option to position help trigger near label for form controls.

## 2.10.0 - 2016-11-04

### Added
- Directive: `sameAs` validator.

## 2.9.3 - 2016-10-14

### Changed
- Expand use of options.lowerCaseOnly in br-input.

## 2.9.2 - 2016-09-13

### Fixed
- Default `Date` fields to null instead of `new Date()`.

## 2.9.1 - 2016-07-26

### Fixed
- Check for form existence when computing validation visibility.

## 2.9.0 - 2016-07-25

### Added
- Enable showValidation by default with br-input.

## 2.8.2 - 2016-07-25

### Fixed
- Use non-selector component when no range options are given.

## 2.8.1 - 2016-07-22

### Fixed
- Always create nodes (expanded form `@id`) for range URL properties.

## 2.8.0 - 2016-07-22

### Added
- Add `lowerCaseOnly` option (true by default) for email brInput types.

## 2.7.0 - 2016-07-20

### Fixed
- Merge queue bugs.

### Changed
- Added library parameter.
- Keep track of property path.
- Lookup schema for values as needed.
- Improve correctness of data while editing.
- Improve field display. (edge cases, objects with ids)
- Index Class types.
- Warn when loading vocab ids with different data.

### Added
- Add alternate to "collapsed". Layout objects can specify "hideHierarchy" and "hideHeader" for fine display control.
- Add "headerStyle" group option.
- Add vocab Displayer type to tie a PropertyGroup to an object type.

## 2.6.1 - 2016-07-14

### Fixed
- Fix help toggle for static form components.

## 2.6.0 - 2016-07-13

### Added
- Add modelTimezone option for date picker.

## 2.5.0 - 2016-06-28

### Added
- Add modelFormat option for date picker.

## 2.4.6 - 2016-06-24

### Fixed
- Properly display form validation errors.

## 2.4.5 - 2016-06-13

### Fixed
- Don't reset library data on each vocab load. Fixes loading of multiple
  vocabs.

### Changed
- Properties and groups from multiple vocabs are merged into one collection in
  the library. Log an info message when the same id is used with identical
  data. Log a warning when the same id has different data.

## 2.4.4 - 2016-06-03

### Fixed
- Static directive template URL typo.

## 2.4.3 - 2016-05-30

### Fixed
- Fix left border radius bug when br-form-control-wrapper is
  not first child but is last.

## 2.4.2 - 2016-05-23

### Fixed
- Fix angular digest iteration bug with form groups.

## 2.4.1 - 2016-05-21

### Fixed
- Remove '{{' and '}}' from br-options expressions for
  br-input, br-textarea, and br-select. It now causes a
  parse error when it was previously required; removal
  should not change old behavior.

## 2.4.0 - 2016-05-19

### Added
- New br-form-control component to provide common functionality
  for form controls like help and validation areas.

### Fixed
- Add missing classes to ui-select in br-select.

## 2.3.4 - 2016-05-10

### Added
- Bedrock template optimization rules.

## 2.3.3 - 2016-05-05

### Fixed
- Remove broken ng-multi-transclude support.

## 2.3.2 - 2016-05-01

### Fixed
- Ensure all prefixed attributes are moved.

## 2.3.1 - 2016-04-22

### Fixed
- Remove invalid `_brInput.legacy` check to re-enable legacy br-input help syntax.

## 2.3.0 - 2016-04-21

### Added
- `<br-form-static>` directive for a properly styled static content form field.

### Removed
- `<br-input>` `static` support. The dual `br-input` and static usage caused
  initialization problems.

## 2.2.0 - 2016-04-21

### Added
- `<br-input>` boolean option `static` that will use a `<p>` vs an `<input>`.
  Transcludes content from `<br-input-static>`.

### Changed
- More usage of `ng-bind-html` for labels.

## 2.1.2 - 2016-04-16

### Fixed
- Ensure inline inputs expand when columns are set.


index.js
/*!
 * Form module.
 *
 * Copyright (c) 2012-2018 Digital Bazaar, Inc. All rights reserved.
 *
 * @author Dave Longley
 */
import angular from 'angular';
import DatepickerComponent from './datepicker-component.js';
import DateFormatDirective from './date-format-directive.js';
import FormDirective from './form-directive.js';
import FormControlComponent from './form-control-component.js';
import FormFieldDirective from './form-field-directive.js';
import FormGroupDirective from './form-group-directive.js';
import FormLibraryService from './form-library-service.js';
import FormStaticDirective from './form-static-directive.js';
import FormUtilsService from './form-utils-service.js';
import HelpToggleDirective from './help-toggle-directive.js';
import InputDirective from './input-directive.js';
import InputManipulatorDirective from './input-manipulator-directive.js';
import InputWatcherDirective from './input-watcher-directive.js';
import RadioGroupDirective from './radio-group-directive.js';
import SelectComponent from './select-component.js';
import TextareaDirective from './textarea-directive.js';
import TrackStateDirective from './track-state-directive.js';
import ValidatorSameAsDirective from './validator-same-as-directive.js';

// TODO: how to make `ngMaterial` optional?
var module = angular.module(
  'bedrock.form', [
    'bedrock.lazyCompile', 'bedrock.resource', 'ngMaterial', 'ngSanitize'
  ]);

module.component('brDatepicker', DatepickerComponent);
module.component('brFormControl', FormControlComponent);
module.component('brSelect', SelectComponent);
module.directive('brInput', InputDirective);
module.directive('brInputManipulator', InputManipulatorDirective);
module.directive('brDateFormat', DateFormatDirective);
module.directive('brForm', FormDirective);
module.directive('brFormField', FormFieldDirective);
module.directive('brFormGroup', FormGroupDirective);
module.directive('brFormStatic', FormStaticDirective);
module.directive('brHelpToggle', HelpToggleDirective);
module.directive('brInputWatcher', InputWatcherDirective);
module.directive('brRadioGroup', RadioGroupDirective);
module.directive('brTextarea', TextareaDirective);
module.directive('brTrackState', TrackStateDirective);
module.directive('brValidatorSameAs', ValidatorSameAsDirective);
module.service('brFormLibraryService', FormLibraryService);
module.service('brFormUtilsService', FormUtilsService);

module.run(function(config) {
  var constants = config.constants = config.constants || {};

  constants.monthNames = [
    'January', 'February', 'March', 'April',
    'May', 'June', 'July', 'August',
    'September', 'October', 'November', 'December'
  ];
  constants.monthNumbers = [
    '01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11',
    '12'
  ];

  // country map sorted by display order
  constants.countries = [
    {code: 'AH', name: 'Afghanistan'},
    {code: 'AL', name: 'Albania'},
    {code: 'DZ', name: 'Algeria'},
    {code: 'AD', name: 'Andorra'},
    {code: 'AO', name: 'Angola'},
    {code: 'AI', name: 'Anguilla'},
    {code: 'AQ', name: 'Antarctica'},
    {code: 'AG', name: 'Antigua'},
    {code: 'AM', name: 'Armenia'},
    {code: 'AR', name: 'Argentina'},
    {code: 'AW', name: 'Aruba'},
    {code: 'AU', name: 'Australia'},
    {code: 'AT', name: 'Austria'},
    {code: 'AZ', name: 'Azerbaidjan'},
    {code: 'BS', name: 'Bahamas'},
    {code: 'BH', name: 'Bahrain'},
    {code: 'BD', name: 'Bangladesh'},
    {code: 'BB', name: 'Barbados'},
    {code: 'BY', name: 'Belarus'},
    {code: 'BE', name: 'Belgium'},
    {code: 'BZ', name: 'Belize'},
    {code: 'BJ', name: 'Benin'},
    {code: 'BM', name: 'Bermuda'},
    {code: 'BT', name: 'Bhutan'},
    {code: 'BA', name: 'Bosnia-Herzegovina'},
    {code: 'BO', name: 'Bolivia'},
    {code: 'BW', name: 'Botswana'},
    {code: 'BV', name: 'Bouvet Island'},
    {code: 'BR', name: 'Brazil'},
    {code: 'BN', name: 'Brunei Darussalam'},
    {code: 'BG', name: 'Bulgaria'},
    {code: 'BF', name: 'Burkina Faso'},
    {code: 'BI', name: 'Burundi'},
    {code: 'KH', name: 'Cambodia'},
    {code: 'CM', name: 'Cameroon'},
    {code: 'CA', name: 'Canada'},
    {code: 'CV', name: 'Cape Verde'},
    {code: 'KY', name: 'Cayman Islands'},
    {code: 'CF', name: 'Central African Republic'},
    {code: 'TD', name: 'Chad'},
    {code: 'CL', name: 'Chile'},
    {code: 'CN', name: 'China'},
    {code: 'CX', name: 'Christmas Island'},
    {code: 'HR', name: 'Croatia'},
    {code: 'CC', name: 'Cocos Islands'},
    {code: 'CO', name: 'Colombia'},
    {code: 'KM', name: 'Comoros'},
    {code: 'CG', name: 'Congo'},
    {code: 'CD', name: 'Congo, Democratic Republic'},
    {code: 'CK', name: 'Cook Islands'},
    {code: 'CR', name: 'Costa Rica'},
    {code: 'CU', name: 'Cuba'},
    {code: 'CY', name: 'Cyprus'},
    {code: 'CZ', name: 'Czech Republic'},
    {code: 'DJ', name: 'Djibouti'},
    {code: 'DK', name: 'Denmark'},
    {code: 'DM', name: 'Dominica'},
    {code: 'DO', name: 'Dominican Republic'},
    {code: 'TP', name: 'East Timor'},
    {code: 'EC', name: 'Ecuador'},
    {code: 'FG', name: 'Egypt'},
    {code: 'SV', name: 'El Salvador'},
    {code: 'ER', name: 'Eritrea'},
    {code: 'EE', name: 'Estonia'},
    {code: 'ET', name: 'Ethiopia'},
    {code: 'GQ', name: 'Equatorial Guinea'},
    {code: 'FK', name: 'Falkland Islands'},
    {code: 'FJ', name: 'Fiji'},
    {code: 'FI', name: 'Finland'},
    {code: 'GF', name: 'French Guyana'},
    {code: 'FO', name: 'Faroe Islands'},
    {code: 'FR', name: 'France'},
    {code: 'GA', name: 'Gabon'},
    {code: 'GM', name: 'Gambia'},
    {code: 'GE', name: 'Georgia'},
    {code: 'DE', name: 'Germany'},
    {code: 'GH', name: 'Ghana'},
    {code: 'GI', name: 'Gibraltar'},
    {code: 'GR', name: 'Greece'},
    {code: 'GL', name: 'Greenland'},
    {code: 'GD', name: 'Grenada'},
    {code: 'GP', name: 'Guadeloupe'},
    {code: 'GU', name: 'Guam'},
    {code: 'GT', name: 'Guatemala'},
    {code: 'GN', name: 'Guinea'},
    {code: 'GW', name: 'Guinea Bissau'},
    {code: 'GY', name: 'Guyana'},
    {code: 'HT', name: 'Haiti'},
    {code: 'HM', name: 'Heard and McDonald Islands'},
    {code: 'HN', name: 'Honduras'},
    {code: 'HK', name: 'Hong Kong'},
    {code: 'HU', name: 'Hungary'},
    {code: 'IS', name: 'Iceland'},
    {code: 'IN', name: 'India'},
    {code: 'ID', name: 'Indonesia'},
    {code: 'IQ', name: 'Iraq'},
    {code: 'IR', name: 'Iran'},
    {code: 'IE', name: 'Ireland'},
    {code: 'IL', name: 'Israel'},
    {code: 'IT', name: 'Italy'},
    {code: 'CI', name: 'Ivory Coast'},
    {code: 'JM', name: 'Jamaica'},
    {code: 'JP', name: 'Japan'},
    {code: 'JO', name: 'Jordan'},
    {code: 'KE', name: 'Kenya'},
    {code: 'KI', name: 'Kiribati'},
    {code: 'KG', name: 'Kyrgyz Republic (Kyrgyzstan)'},
    {code: 'KW', name: 'Kuwait'},
    {code: 'KZ', name: 'Kazakhstan'},
    {code: 'LA', name: 'Laos'},
    {code: 'LV', name: 'Latvia'},
    {code: 'LB', name: 'Lebanon'},
    {code: 'LS', name: 'Lesotho'},
    {code: 'LR', name: 'Liberia'},
    {code: 'LI', name: 'Liechtenstein'},
    {code: 'LT', name: 'Lithuania'},
    {code: 'LY', name: 'Libya'},
    {code: 'LU', name: 'Luxembourg'},
    {code: 'MO', name: 'Macau'},
    {code: 'MK', name: 'Macedonia'},
    {code: 'MG', name: 'Madagascar'},
    {code: 'ML', name: 'Mali'},
    {code: 'MT', name: 'Malta'},
    {code: 'MV', name: 'Maldives'},
    {code: 'MW', name: 'Malawi'},
    {code: 'MY', name: 'Malaysia'},
    {code: 'MH', name: 'Marshall Islands'},
    {code: 'MQ', name: 'Martinique'},
    {code: 'MR', name: 'Mauritania'},
    {code: 'MU', name: 'Mauritius'},
    {code: 'YT', name: 'Mayotte'},
    {code: 'MX', name: 'Mexico'},
    {code: 'FM', name: 'Micronesia'},
    {code: 'MD', name: 'Moldavia'},
    {code: 'MC', name: 'Monaco'},
    {code: 'MN', name: 'Mongolia'},
    {code: 'MS', name: 'Montserrat'},
    {code: 'MA', name: 'Morocco'},
    {code: 'MZ', name: 'Mozambique'},
    {code: 'MM', name: 'Myanmar'},
    {code: 'NA', name: 'Namibia'},
    {code: 'NR', name: 'Nauru'},
    {code: 'NP', name: 'Nepal'},
    {code: 'NL', name: 'Netherlands'},
    {code: 'NC', name: 'New Caledonia'},
    {code: 'NZ', name: 'New Zealand'},
    {code: 'NE', name: 'Niger'},
    {code: 'NG', name: 'Nigeria'},
    {code: 'NI', name: 'Nicaragua'},
    {code: 'NU', name: 'Niue'},
    {code: 'NF', name: 'Norfolk Island'},
    {code: 'KP', name: 'North Korea'},
    {code: 'MP', name: 'Northern Mariana Islands'},
    {code: 'NO', name: 'Norway'},
    {code: 'OM', name: 'Oman'},
    {code: 'PW', name: 'Palau'},
    {code: 'PK', name: 'Pakistan'},
    {code: 'PA', name: 'Panama'},
    {code: 'PG', name: 'Papua New Guinea'},
    {code: 'PY', name: 'Paraguay'},
    {code: 'PE', name: 'Peru'},
    {code: 'PH', name: 'Philippines'},
    {code: 'PN', name: 'Pitcairn Island'},
    {code: 'PR', name: 'Puerto Rico'},
    {code: 'PL', name: 'Poland'},
    {code: 'PF', name: 'Polynesia'},
    {code: 'PT', name: 'Portugal'},
    {code: 'QA', name: 'Qatar'},
    {code: 'RE', name: 'Reunion'},
    {code: 'RO', name: 'Romania'},
    {code: 'RU', name: 'Russia'},
    {code: 'RW', name: 'Rwanda'},
    {code: 'SH', name: 'Saint Helena'},
    {code: 'KN', name: 'Saint Kitts and Nevis Anguilla'},
    {code: 'LC', name: 'Saint Lucia'},
    {code: 'PM', name: 'Saint Pierre and Miquelon'},
    {code: 'ST', name: 'Saint Tome and Principe'},
    {code: 'WS', name: 'Samoa'},
    {code: 'SM', name: 'San Marino'},
    {code: 'SA', name: 'Saudi Arabia'},
    {code: 'SN', name: 'Senegal'},
    {code: 'SC', name: 'Seychelles'},
    {code: 'SL', name: 'Sierra Leone'},
    {code: 'SG', name: 'Singapore'},
    {code: 'SK', name: 'Slovak Republic'},
    {code: 'SI', name: 'Slovenia'},
    {code: 'SB', name: 'Solomon Islands'},
    {code: 'SO', name: 'Somalia'},
    {code: 'ZA', name: 'South Africa'},
    {code: 'GS', name: 'S. Georgia and S. Sandwich Isls.'},
    {code: 'KR', name: 'South Korea'},
    {code: 'ES', name: 'Spain'},
    {code: 'LK', name: 'Sri Lanka'},
    {code: 'SD', name: 'Sudan'},
    {code: 'SS', name: 'South Sudan'},
    {code: 'SR', name: 'Suriname'},
    {code: 'SJ', name: 'Svalbard and Jan Mayen Islands'},
    {code: 'SZ', name: 'Swaziland'},
    {code: 'SE', name: 'Sweden'},
    {code: 'CH', name: 'Switzerland'},
    {code: 'SY', name: 'Syria'},
    {code: 'TJ', name: 'Tadjikistan'},
    {code: 'TW', name: 'Taiwan'},
    {code: 'TZ', name: 'Tanzania'},
    {code: 'TG', name: 'Togo'},
    {code: 'TK', name: 'Tokelau'},
    {code: 'TO', name: 'Tonga'},
    {code: 'TH', name: 'Thailand'},
    {code: 'TN', name: 'Tunisia'},
    {code: 'TR', name: 'Turkey'},
    {code: 'TM', name: 'Turkmenistan'},
    {code: 'TC', name: 'Turks and Caicos Islands'},
    {code: 'TT', name: 'Trinidad and Tobago'},
    {code: 'TV', name: 'Tuvalu'},
    {code: 'UA', name: 'Ukraine'},
    {code: 'UG', name: 'Uganda'},
    {code: 'AE', name: 'United Arab Emirates'},
    {code: 'UK', name: 'United Kingdom'},
    {code: 'US', name: 'United States'},
    {code: 'UY', name: 'Uruguay'},
    {code: 'UZ', name: 'Uzbekistan'},
    {code: 'VA', name: 'Vatican City'},
    {code: 'VC', name: 'Saint Vincent and Grenadines'},
    {code: 'VU', name: 'Vanuatu'},
    {code: 'VE', name: 'Venezuela'},
    {code: 'VG', name: 'Virgin Islands (British)'},
    {code: 'VI', name: 'Virgin Islands (USA)'},
    {code: 'VN', name: 'Vietnam'},
    {code: 'WF', name: 'Wallis and Futuna Islands'},
    {code: 'EH', name: 'Western Sahara'},
    {code: 'YE', name: 'Yemen'},
    {code: 'YU', name: 'Yugoslavia'},
    {code: 'ZR', name: 'Zaire'},
    {code: 'ZM', name: 'Zambia'},
    {code: 'ZW', name: 'Zimbabwe'}
  ];

  // create {index: #, label: "01 - January"} labels
  constants.monthLabels = [];
  for(var i = 0; i < 12; ++i) {
    constants.monthLabels[i] = {
      index: i + 1,
      label: constants.monthNumbers[i] + ' - ' + constants.monthNames[i]
    };
  }

  // next ten 10 years for expiration dates (quick hack)
  constants.years = [];
  var year = new Date().getFullYear();
  for(var i = 0; i < 10; ++i) {
    constants.years.push(year + i);
  }
});
